npm/extensions.bzl

```diff
@@ -28,45 +27,61 @@ def _extension_impl(module_ctx):
             registries = {}
             lock_importers, lock_packages = utils.parse_pnpm_lock(module_ctx.read(attr.pnpm_lock))
             importers, packages = translate_to_transitive_closure(lock_importers, lock_packages, attr.prod, attr.dev, attr.no_optional)
-            imports = npm_translate_lock_helpers.gen_npm_imports(importers, packages, attr.pnpm_lock.package, npm_translate_lock_name, attr, registries, utils.default_registry())
+            imports = npm_translate_lock_helpers.gen_npm_imports(importers, packages, attr.pnpm_lock.package, attr.name, attr, registries, utils.default_registry())
             for i in imports:
                 npm_import(
                     name = i.name,
-                    package = i.package,
-                    version = i.version,
-                    link_packages = i.link_packages,
                     custom_postinstall = i.custom_postinstall,
                     deps = i.deps,
                     integrity = i.integrity,
+                    lifecycle_hooks = i.lifecycle_hooks,
+                    link_packages = i.link_packages,
+                    npm_translate_lock_repo = attr.name,
+                    package = i.package,
                     patch_args = i.patch_args,
                     patches = i.patches,
                     root_package = i.root_package,
-                    lifecycle_hooks = i.lifecycle_hooks,
                     transitive_closure = i.transitive_closure,
                     url = i.url,
-                    npm_translate_lock_repo = "npm",
+                    version = i.version,
                 )
 
         for i in mod.tags.npm_import:
             npm_import(
                 name = i.name,
-                package = i.package,
-                version = i.version,
-                link_packages = i.link_packages,
+                custom_postinstall = i.custom_postinstall,
                 integrity = i.integrity,
-                patch_args = i.patch_args,
-                patches = i.patches,
                 lifecycle_hooks = i.lifecycle_hooks,
-                custom_postinstall = i.custom_postinstall,
+                link_packages = i.link_packages,
                 link_workspace = i.link_workspace,
-                url = i.url,
+                package = i.package,
+                patch_args = i.patch_args,
+                patches = i.patches,
                 root_package = i.root_package,
+                url = i.url,
+                version = i.version,
             )
 
+def _npm_translate_lock_attrs():
+    attrs = dict(**npm_translate_lock_lib.attrs)
+
+    # Add macro attrs that aren't in the rule attrs.
+    attrs["name"] = attr.string()
+
```

Comment:
Not much here in the prefactor, but there are several attributes I will need to add here and below from the other PR.

Reply:
Since we invoke the macro, we won't get all of the attributes we need from `**npm_translate_lock_lib.attrs`, which just includes the rule attributes. There are a few macro parameters that we need to represent as extension tag class attributes for bzlmod.

Reply:
👍